src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -316,40 +333,141 @@ public void insertBufferedRows() {
+   * ingestion and introduce lags in committing offsets to Kafka.
+   *
+   * <p>Note that insertRows API does perform channel validation which might throw SFException if
+   * channel is invalidated.
+   *
+   * <p>It can also send errors {@link
+   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
+   * {@link InsertValidationResponse}
+   *
+   * @param buffer buffer to insert into snowflake
+   * @return InsertValidationResponse object sent from insertRows API.
+   */
+  private InsertValidationResponse insertRowsWithFallback(StreamingBuffer buffer) {
+    Fallback<Object> reopenChannelFallbackExecutorForInsertRows =
+        Fallback.builder(() -> insertRowsFallbackSupplier(buffer))
+            .handle(SFException.class)
+            .onFailure(
+                event ->
+                    LOGGER.error(
+                        String.format(
+                            "%s Failed to open Channel or fetching offsetToken for channel:%s",
+                            StreamingApiFallbackInvoker.INSERT_ROWS_FALLBACK,
+                            this.getChannelName()),
+                        event.getException()))
+            .build();
 
-    // should we retry if response.hasErrors() == true?
-    // how can we retry if buffer was cleared?
-    // lets have an exponential back off and jitter retry policy
+    return Failsafe.with(reopenChannelFallbackExecutorForInsertRows)
+        .get(new InsertRowsApiResponseSupplier(this.channel, buffer));
+  }
 
-    if (response.hasErrors()) {
-      LOGGER.error(
-          "Inserted rows API response has errors for rows with offset start:{} and end:{}",
-          intermediateBuffer.getFirstOffset(),
-          intermediateBuffer.getLastOffset());
-      response
-          .getInsertErrors()
-          .forEach(
-              insertError -> {
-                LOGGER.error(
-                    "Insert row Error message:{}, with ex:{}",
-                    insertError.getMessage(),
-                    insertError.getException().getMessage());
-              });
+  /** Invokes the API given the channel and streaming Buffer. */
+  private static class InsertRowsApiResponseSupplier
+      implements CheckedSupplier<InsertValidationResponse> {
+
+    private final SnowflakeStreamingIngestChannel channel;
+
+    private final StreamingBuffer streamingBuffer;
+
+    private InsertRowsApiResponseSupplier(
+        SnowflakeStreamingIngestChannel channel, StreamingBuffer streamingBuffer) {
+      this.channel = channel;
+      this.streamingBuffer = streamingBuffer;
+    }
+
+    @Override
+    public InsertValidationResponse get() throws Throwable {
+      LOGGER.debug(
+          "Invoking insertRows API for channel:{}, noOfRecords:{}, startOffset:{}, endOffset:{}",
+          this.channel.getFullyQualifiedName(),
+          streamingBuffer.getNumOfRecords(),
+          streamingBuffer.getFirstOffset(),
+          streamingBuffer.getLastOffset());
```

Comment:
nit: use the same pattern
```suggestion
          this.channel.getFullyQualifiedName(),
          this.streamingBuffer.getNumOfRecords(),
          this.streamingBuffer.getFirstOffset(),
          this.streamingBuffer.getLastOffset());
```

Reply:
Oops, I think the rebase was mysterious.